### PR TITLE
Selenium nodechrome 3765

### PR DESCRIPTION
**Pull request by utieyin**
* docker-selenium node-firefox addition.  In relation to https://github.com/orgs/chainguard-dev/projects/45/views/31?pane=issue&itemId=70997726&issue=chainguard-dev%7Cimage-requests%7C3765

An example config and several subpackages now exist for creating multiple components of docker-selenium
#### For new package PRs only

- [x] REQUIRED - The package is available under an OSI-approved or FSF-approved license
- [x] REQUIRED - The version of the package is still receiving security updates

### CODE REVIEW

File: docker-selenium.yaml [hunk 1]
```diff
@@ -102,48 +102,81 @@ pipeline:
...
 
 subpackages:
-  - name: docker-selenium-supervisor-config
-    description: Docker Selenium supervisor configuration
-    dependencies:
-      replaces:
-        - supervisor-config
-      provides:
-        - supervisor-config
+  - name: ${{package.name}}-config
```
Comment (Dentrax) [inline docker-selenium.yaml:105]: We should bump the epoch above

File: docker-selenium.yaml [hunk 2]
```diff
@@ -173,42 +205,133 @@ subpackages:
...
-      - working-directory: Standalone
+              mkdir -p ${{targets.contextdir}}/opt/bin/
+              cp wrap_chrome_binary ${{targets.contextdir}}/opt/bin/wrap_chrome_binary
+    test:
+      environment:
+        contents:
+          packages:
+      pipeline:
+        - runs: |
+            /opt/bin/wrap_chrome_binary &
```
Comment (Dentrax) [inline docker-selenium.yaml:277]: should we use `test/daemon-check-output` here?

Comment (utieyin) [reply docker-selenium.yaml:277]: yep, doing that currently. Working through all the packages

File: docker-selenium.yaml (hunk 1 above)
Comment (utieyin) [reply docker-selenium.yaml:105]: done

File: docker-selenium.yaml [hunk 3]
```diff
@@ -111,39 +117,108 @@ subpackages:
...
+        - selenium-server-standalone
+        - ${{package.name}}
+        - ${{package.name}}-config
+    pipeline:
+      - working-directory: Standalone
+        pipeline:
+          - runs: |
+              mkdir -p ${{targets.contextdir}}/etc/supervisor/conf.d
+              mkdir -p ${{targets.contextdir}}/opt/bin
+              cp start-selenium-standalone.sh ${{targets.contextdir}}/opt/bin/start-selenium-standalone.sh
```
Comment (maxgio92) [inline docker-selenium.yaml:141]: Side of side comment: you can omit the target filename `start-selenium-standalone.sh` :-)

File: docker-selenium.yaml [hunk 4]
```diff
@@ -173,42 +246,137 @@ subpackages:
...
+        - name: "Check selenium node-firefox status status"
+          uses: test/daemon-check-output
+          with:
+            start: "/opt/bin/start-selenium-node.sh"
+            timeout: 60
+            expected_output: |
+              Adding Firefox for
+              Started Selenium node
+
+  - name: ${{package.name}}-standalone-chrome
```
Comment (EyeCantCU) [inline docker-selenium.yaml:370]: I think we can drop the meta package but otherwise, this looks great!

Comment (utieyin) [reply docker-selenium.yaml]: it was agreed to keep this in for clarity

File: docker-selenium.yaml [hunk 5]
```diff
@@ -173,42 +246,464 @@ subpackages:
...
+        - name: "Check selenium node-base status status"
+          uses: test/daemon-check-output
+          with:
+            start: "/opt/bin/start-selenium-node.sh"
+            timeout: 60
+            expected_output: |
+              Started Selenium node
+
+  # https://github.com/SeleniumHQ/docker-selenium/blob/trunk/NodeChrome/Dockerfile
+  - name: ${{package.name}}-node-chrome
```
Comment (EyeCantCU) [inline docker-selenium.yaml:301]: This is a duplicate of `-standalone-chromium` below. We should remove this and use that

**Review by EyeCantCU — CHANGES_REQUESTED** [review]
One last change, otherwise looks good

File: docker-selenium.yaml (hunk 5 above)
Comment (utieyin) [reply docker-selenium.yaml]: chrome package removed

File: docker-selenium.yaml [hunk 6]
```diff
@@ -173,42 +246,432 @@ subpackages:
...
+        - name: "Check selenium distributor status"
+          uses: test/daemon-check-output
+          with:
+            start: "/opt/bin/start-selenium-grid-distributor.sh"
+            timeout: 60
+            expected_output: |
+              Started Selenium Distributor
+
+  - name: ${{package.name}}-node-docker
+    description: Docker Selenium noded-ocker
```
Comment (EyeCantCU) [inline docker-selenium.yaml:468]:
```suggestion
    description: Docker Selenium noded-docker
```

**Review by EyeCantCU — CHANGES_REQUESTED** [review]

File: docker-selenium.yaml [hunk 7]
```diff
@@ -173,42 +246,432 @@ subpackages:
...
+      pipeline:
+        - name: "Check selenium router status"
+          uses: test/daemon-check-output
+          with:
+            start: "/opt/bin/start-selenium-grid-router.sh"
+            timeout: 60
+            expected_output: |
+              Started Selenium Router
+
+  - name: ${{package.name}}-session-queueeue
```
Comment (EyeCantCU) [inline docker-selenium.yaml:548]: Typo

**Review by EyeCantCU — DISMISSED** [review]

File: docker-selenium.yaml (hunk 7 above)
Comment (utieyin) [reply docker-selenium.yaml]: fixed

**Review by Dentrax — COMMENTED** [review]
Great work and nice tests! Dropped a few comments.

File: docker-selenium.yaml [hunk 8]
```diff
@@ -173,42 +265,425 @@ subpackages:
...
+              mkdir -p ${{targets.contextdir}}/tmp/.X11-unix
+              mkdir -p ${{targets.contextdir}}/home/$SEL_USER/.vnc
+              mkdir -p ${{targets.contextdir}}/opt/selenium
+              mkdir -p ${{targets.contextdir}}/opt/selenium
+              echo "${SEL_PASSWD}" | x11vnc -storepasswd - ${{targets.contextdir}}/home/$SEL_USER/.vnc/passwd
+    test:
+      environment:
+        contents:
+          packages:
+            - geckodriver
```
[the remark this answers is not in the document]
Comment (utieyin) [reply docker-selenium.yaml:300]: @Dentrax , the nodebase component which is been tested here acts as a base image to build node-chrome/firefox etc. The geckodriver package just lets us test what would be the outcome if nodebase was used as base package for nodefirefox in this case. It is not a runtime dep on nodebase per se

File: docker-selenium.yaml [hunk 9]
```diff
@@ -111,39 +139,104 @@ subpackages:
...
+          SE_RELAX_CHECKS: true
+          SE_REJECT_UNSUPPORTED_CAPS: true
+          SE_OTEL_SERVICE_NAME: "selenium-standalone"
+          SE_NODE_SESSION_TIMEOUT: 150
+          SE_NODE_OVERRIDE_MAX_SESSIONS: 10
+          DRAIN_AFTER_SESSION_COUNT: 10
+          SE_DRAIN_AFTER_SESSION_COUNT: 4
+          SE_NODE_MAX_SESSIONS: 10
+          SE_NODE_BROWSER_NAME: chrome
+          SE_NODE_BROWSER_VERSION: $(chromium --version | cut -d ' ' -f 2)
```
[the remark this answers is not in the document]
Comment (utieyin) [reply docker-selenium.yaml:186]: @Dentrax it does look like melange supports it as I was able to successfully execute the test which fails without the browser version set. I can set a static value here if that is preferred. Just thought this approach lets us use any version of chromium

File: docker-selenium.yaml [hunk 10]
```diff
@@ -67,42 +62,75 @@ pipeline:
...
-          mkdir -p ${{targets.destdir}}/external_jars
-          curl -sSLfO https://github.com/coursier/launchers/raw/master/coursier
-          chmod +x coursier
-          # The netty-codec-http bump is for CVE-2024-29025
-          ./coursier fetch --classpath --cache ${{targets.destdir}}/external_jars \
-            io.opentelemetry:opentelemetry-exporter-otlp:${OPENTELEMETRY_VERSION} \
-            io.netty:netty-codec-http:4.1.108.Final \
-            io.grpc:grpc-netty:${GRPC_VERSION} > ${{targets.destdir}}/external_jars/.classpath.txt
+              mkdir -p ${{targets.contextdir}}/external_jars
+              curl -sSLfO https://github.com/coursier/launchers/raw/master/coursier
```
[the remark this answers is not in the document]
Comment (utieyin) [reply docker-selenium.yaml:102]: @Dentrax considering that the scope of this has crept up gradually, I can create an issue for the coursier package separately

File: docker-selenium.yaml [hunk 11]
```diff
@@ -18,11 +18,6 @@ package:
       - bash
       - busybox
       - coreutils
-      - openjdk-11
-      - selenium-server-compat
-      - sudo-rs
-      - supervisor
-      - tzdata
```
[the remark this answers is not in the document]
Comment (utieyin) [reply docker-selenium.yaml:25]: There will no longer be a docker-selenium image. We are now packaging the components just as upstream docker-selenium does. Here is the PR for that work https://github.com/chainguard-images/images-private/pull/5681

File: docker-selenium.yaml [hunk 12]
```diff
@@ -174,42 +266,425 @@ subpackages:
...
+        environment:
+          SE_EVENT_BUS_PUBLISH_PORT: 4442
+          SE_EVENT_BUS_SUBSCRIBE_PORT: 4443
+          SE_EVENT_BUS_HOST: selenium-hub
+          CONFIG_FILE: /opt/selenium/config.toml
+          SE_BIND_HOST: true
+      pipeline:
+        - runs: |
+            firefox --version
+        - name: "Check selenium node-firefox status status"
```
Comment (maxgio92) [inline docker-selenium.yaml:382]: Typo here and similar with "status status"

**Comment by maxgio92** [comment]
Great work!

**Review by maxgio92 — APPROVED** [review]

**Pull request merged**